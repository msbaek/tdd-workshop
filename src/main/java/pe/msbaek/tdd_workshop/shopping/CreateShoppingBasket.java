package pe.msbaek.tdd_workshop.shopping;

import org.springframework.web.bind.annotation.*;

import java.math.BigDecimal;
import java.util.List;

@RestController
@RequestMapping("/api/baskets")
public class CreateShoppingBasket {

    private final BasketRepository basketRepository;

    public CreateShoppingBasket(BasketRepository basketRepository) {
        this.basketRepository = basketRepository;
    }

    @PostMapping
    public BasketResponse createBasket(@RequestBody BasketItemRequests requests) {
        // 빈 장바구니 검증
        if (requests.items() == null || requests.items().isEmpty()) {
            throw new IllegalArgumentException("장바구니가 비어 있어서 영수증을 생성할 수 없습니다.");
        }
        
        // Fake it: 하드코딩으로 테스트만 성공시키기
        List<BasketItem> items = requests.items().stream()
                .map(req -> new BasketItem(req.name(), req.price(), req.quantity()))
                .toList();
        
        Basket basket = new Basket(null, items);
        Basket savedBasket = basketRepository.save(basket);
        
        return new BasketResponse(savedBasket.getId().toString());
    }

    @GetMapping("/{basketId}")
    public BasketDetailsResponse getBasketDetails(@PathVariable String basketId) {
        Long id = Long.valueOf(basketId);
        Basket basket = basketRepository.findById(id).orElseThrow();
        
        // 실제 계산으로 변경
        List<BasketItemDto> itemDtos = basket.getItems().stream()
                .map(item -> new BasketItemDto(
                        item.getName(),
                        item.getPrice(),
                        item.getQuantity(),
                        item.getPrice().multiply(BigDecimal.valueOf(item.getQuantity()))
                ))
                .toList();
        
        // 소계 계산
        BigDecimal subtotal = itemDtos.stream()
                .map(BasketItemDto::total)
                .reduce(BigDecimal.ZERO, BigDecimal::add);
        
        // 할인 계산 (10,000원 이하는 할인 없음)
        BigDecimal discount = BigDecimal.ZERO;
        String discountRate = "할인 없음";
        
        // 최종 금액
        BigDecimal total = subtotal.subtract(discount);
        
        return new BasketDetailsResponse(
                basketId,
                itemDtos,
                subtotal,
                discount,
                total,
                discountRate
        );
    }

    // Inner records for request/response
    public record BasketItemRequests(List<BasketItemRequest> items) {}
    public record BasketItemRequest(String name, BigDecimal price, int quantity) {}
    public record BasketResponse(String basketId) {}
    public record BasketDetailsResponse(
            String basketId,
            List<BasketItemDto> items,
            BigDecimal subtotal,
            BigDecimal discount,
            BigDecimal total,
            String discountRate
    ) {}
    public record BasketItemDto(String name, BigDecimal price, int quantity, BigDecimal total) {}
}
