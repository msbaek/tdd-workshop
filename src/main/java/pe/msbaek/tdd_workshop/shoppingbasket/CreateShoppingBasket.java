package pe.msbaek.tdd_workshop.shoppingbasket;

import java.math.BigDecimal;
import java.util.List;

public class CreateShoppingBasket {

    // Request/Response DTOs
    record BasketItemRequests(List<BasketItemRequest> items) {}
    record BasketItemRequest(String name, BigDecimal price, int quantity) {}
    record BasketResponse(String basketId) {}
    record BasketDetailsResponse(String basketId, List<BasketItem> items,
                                 BigDecimal subtotal, BigDecimal discount, BigDecimal total) {}
    record BasketItem(String name, BigDecimal price, int quantity, BigDecimal itemTotal) {}
}
