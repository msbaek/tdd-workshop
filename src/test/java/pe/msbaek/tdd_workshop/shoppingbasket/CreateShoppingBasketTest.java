package pe.msbaek.tdd_workshop.shoppingbasket;

import com.fasterxml.jackson.databind.ObjectMapper;
import org.approvaltests.Approvals;
import org.junit.jupiter.api.Disabled;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.autoconfigure.web.servlet.AutoConfigureMockMvc;
import org.springframework.boot.test.context.SpringBootTest;
import org.springframework.http.MediaType;
import org.springframework.test.web.servlet.MockMvc;
import org.springframework.test.web.servlet.MvcResult;
import pe.msbaek.tdd_workshop.shoppingbasket.CreateShoppingBasket.BasketDetailsResponse;
import pe.msbaek.tdd_workshop.shoppingbasket.CreateShoppingBasket.BasketItemRequest;
import pe.msbaek.tdd_workshop.shoppingbasket.CreateShoppingBasket.BasketItemRequests;
import pe.msbaek.tdd_workshop.shoppingbasket.CreateShoppingBasket.BasketResponse;

import java.math.BigDecimal;
import java.util.List;

import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.get;
import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.post;
import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.status;

@SpringBootTest
@AutoConfigureMockMvc
public class CreateShoppingBasketTest {
    @Autowired
    private MockMvc mockMvc;

    @Autowired
    private ObjectMapper objectMapper;

    @Disabled("아직 기능 구현이 완료되지 않았습니다.")
    @DisplayName("여러 상품이 있고 20,000원 초과 시 10% 할인 적용되는 청구서 생성")
    @Test
    void create_and_verify_basket() throws Exception {
        // given
        // 장바구니에 상품 일괄 추가
        BasketItemRequests items = new BasketItemRequests(List.of(
                new BasketItemRequest("스마트폰 케이스", BigDecimal.valueOf(15000), 1),
                new BasketItemRequest("보호필름", BigDecimal.valueOf(5000), 2),
                new BasketItemRequest("충전 케이블", BigDecimal.valueOf(8000), 1)
        ));

        // when
        MvcResult postResult = mockMvc.perform(post("/api/baskets")
                        .contentType(MediaType.APPLICATION_JSON)
                        .content(objectMapper.writeValueAsString(items)))
                .andExpect(status().isOk())
                .andReturn();

        BasketResponse response = objectMapper.readValue(
                postResult.getResponse().getContentAsString(),
                BasketResponse.class);

        // 생성된 장바구니 id 획득
        String basketId = response.basketId();

        // assert: get을 통해 같은 api 레벨에서 결과 확인
        MvcResult getResult = mockMvc.perform(get("/api/baskets/" + basketId))
                .andExpect(status().isOk())
                .andReturn();

        BasketDetailsResponse basketDetails = objectMapper.readValue(
                getResult.getResponse().getContentAsString(),
                BasketDetailsResponse.class);

        // 응답 내용 검증
        Approvals.verify(printBasketDetails(basketDetails));
    }

    /**
     * 영수증을 출력하는 메소드
     */
    private String printBasketDetails(BasketDetailsResponse basketDetails) {
        return """
                ===== 영수증 =====
                품목:
                - 스마트폰 케이스 1개 (단가: 15,000원, 총액: 15,000원)
                - 보호필름 2개 (단가: 5,000원, 총액: 10,000원)
                - 충전 케이블 1개 (단가: 8,000원, 총액: 8,000원)
                소계: 33,000원
                할인: 3,300원 (10% 할인)
                최종 결제 금액: 29,700원
                ==================
                """;
    }
}