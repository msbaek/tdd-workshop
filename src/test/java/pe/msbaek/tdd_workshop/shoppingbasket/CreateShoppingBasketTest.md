# CreateShoppingBasket TDD 구현

## 전체적인 절차

## 1. **SRS(소프트웨어 요구사항 명세서) 작성**

### CreateShoppingBasket 규칙 및 요구사항

- 규칙
    - 기본 규칙
        - 사용자는 장바구니에 상품을 추가할 수 있음
        - 각 상품은 이름, 단가, 수량을 가짐
        - 총 가격은 `단가 * 수량`의 합계임
    - 예외 처리 규칙
        - 장바구니가 비어 있으면 청구서를 생성할 수 없음
        - 빈 장바구니에서 영수증 생성을 요청하면 시스템은 적절한 오류 메시지와 함께 예외를 발생시켜야 함
    - 할인 정책
        - 총 금액이 20,000원 이상이면 10% 할인 제공
        - 총 금액이 10,000원 초과 20,000원 미만이면 5% 할인 제공
        - 총 금액이 10,000원 이하면 할인 없음
- 청구서 생성 요구사항
    - 기본 청구서 정보
        - 시스템은 장바구니의 각 상품 정보(이름, 단가, 수량, 총액)를 표시해야 함
        - 소계(할인 전 총액)를 계산하고 표시해야 함
        - 적용된 할인 금액과 할인율을 표시해야 함
        - 최종 결제 금액(할인 후 금액)을 계산하고 표시해야 함
    - 청구서 형식
        - 영수증 형태로 품목별 상세 정보를 나열
        - 소계, 할인, 최종 결제 금액을 명확히 구분하여 표시
        - 할인이 적용된 경우 할인율을 함께 표시

## 2. **SRS를 잘 설명할 수 있는 예제 목록 작성**

### CreateShoppingBasket 예제

- 예제1: 빈 장바구니 - 청구서 발행 불가
    - 장바구니에 상품이 없는 상태에서 영수증 생성 요청
    - 기대 결과: 시스템은 적절한 오류 메시지와 함께 예외를 발생시켜야 함

- 예제2: 10,000원 이하 - 할인 없음
    - 보호필름 2개 (단가: 5,000원, 총액: 10,000원)
    - 소계: 10,000원
    - 할인: 0원 (할인 없음)
    - 최종 결제 금액: 10,000원

- 예제3: 10,000원 초과 20,000원 미만 - 5% 할인 적용
    - 스마트폰 케이스 1개 (단가: 12,000원, 총액: 12,000원)
    - 보호필름 1개 (단가: 5,000원, 총액: 5,000원)
    - 소계: 17,000원
    - 할인: 850원 (5% 할인)
    - 최종 결제 금액: 16,150원

- 예제4: 정확히 20,000원 - 10% 할인 적용
    - 스마트폰 케이스 1개 (단가: 15,000원, 총액: 15,000원)
    - 보호필름 1개 (단가: 5,000원, 총액: 5,000원)
    - 소계: 20,000원
    - 할인: 2,000원 (10% 할인)
    - 최종 결제 금액: 18,000원

- 예제5: 20,000원 초과 - 10% 할인 적용 (복합 예제)
    - 스마트폰 케이스 1개 (단가: 15,000원, 총액: 15,000원)
    - 보호필름 2개 (단가: 5,000원, 총액: 10,000원)
    - 충전 케이블 1개 (단가: 8,000원, 총액: 8,000원)
    - 소계: 33,000원
    - 할인: 3,300원 (10% 할인)
    - 최종 결제 금액: 29,700원

## 3. **High Level Test 작성**

### 대표 예제 선택
- 선택된 예제: 예제5 (20,000원 초과 - 10% 할인 적용)
- 선택 이유: 여러 상품, 수량 처리, 할인 계산 등 가장 많은 제약 조건을 포함하는 일반적인 케이스

### High Level Test 구현

```mermaid
classDiagram
    class Basket {
        +Long id
        +List~BasketItem~ items
    }
    
    class BasketItem {
        +String name
        +BigDecimal price
        +int quantity
        +BigDecimal itemTotal
    }
    
    Basket --> BasketItem
```

## 4. **테스트 케이스 목록 작성**

## 5. **Walking Skeleton 구현**

## 6. **테스트 리스트에서 테스트 선택해서 테스트 추가하기(더 이상 추가할 테스트가 없을때까지)**

'<implement-each-test-rule>' 준수

## 7. **High Level Test 활성화**

## 8. **Jpa Repository 구현**

## 진행 상황

현재 단계: 3단계 (High Level Test 작성) 완료
