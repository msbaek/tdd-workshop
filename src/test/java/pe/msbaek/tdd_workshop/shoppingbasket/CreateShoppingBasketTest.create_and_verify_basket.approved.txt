===== 영수증 =====
품목:
- 스마트폰 케이스 1개 (단가: 15,000원, 총액: 15,000원)
- 보호필름 2개 (단가: 5,000원, 총액: 10,000원)
- 충전 케이블 1개 (단가: 8,000원, 총액: 8,000원)
소계: 33,000원
할인: 3,300원 (10% 할인)
최종 결제 금액: 29,700원
==================
